### PR TITLE
Add debug shape to invisible components.

### PR DESCRIPTION
**Pull request by EternalWind**
I didn't use the models you given to me because their sizes are not uniform
and I don't know how to modify them in blender. So I havn't add the artists
to credit yet.

### CODE REVIEW

**Comment by EternalWind** [comment]
Svenstaro: If I make the Component do the stuff MeshComponent do to implement the debug shape, then I have to add Component::OnUpdate() to all the subclass's OnUpdate() method. The same to Component::OnCreate(), Component::OnDestroy(). (Well, I can move the stuff in Component::OnCreate() to Component::Create() though and then I don't need to add Component::OnCreate() to the subclass's OnCreate() method. However, the rest still need to do that.) What do you think?

File: engine/src/Scene/Component.cpp [hunk 1]
```diff
@@ -75,6 +82,7 @@ void Component::Destroy() {
         Disable();
         emit ComponentDestroyed();
         OnDestroy();
+		_DestroyDebugMesh();
```
Comment (svenstaro) [inline engine/src/Scene/Component.cpp]: Wrong indentation? Here as well as in many other places.

File: engine/src/Scene/Component.cpp [hunk 2]
```diff
@@ -12,14 +12,19 @@
...
 namespace dt {
 
-Component::Component(const QString& name)
+Component::Component(const QString& name, const QString& mesh_handle_d, const QString& mesh_component_name_d)
     : mName(name),
       mIsEnabled(false),
-      mIsCreated(false) {
+      mIsCreated(false),
+      mMeshHandle(mesh_handle_d),
+      mMeshComponentName(mesh_component_name_d){
```
Comment (svenstaro) [inline engine/src/Scene/Component.cpp]: Space between ) and {

File: engine/src/Scene/Component.cpp [hunk 3]
```diff
@@ -102,4 +110,43 @@ bool Component::IsEnabled() {
...
+void Component::HideDebug() {
+	mEntity->setVisible(false);
+}
+
+void Component::ShowDebug() {
+	mEntity->setVisible(true);
+}
+
+void Component::_LoadDebugMesh() {
+	//_DestroyDebugMesh();
```
Comment (svenstaro) [inline engine/src/Scene/Component.cpp]: Why the comment?

File: engine/src/Scene/Node.cpp [hunk 4]
```diff
@@ -223,7 +223,7 @@ void Node::_UpdateAllComponents(double time_diff) {
 
     for(auto iter = mComponents.begin(); iter != mComponents.end(); ++iter) {
         if(iter->second->IsEnabled()) {
-            iter->second->OnUpdate(time_diff);
+			iter->second->Update(time_diff);
```
Comment (svenstaro) [inline engine/src/Scene/Node.cpp]: Indentation

File: engine/src/Scene/Component.cpp [hunk 5]
```diff
@@ -102,4 +112,43 @@ bool Component::IsEnabled() {
...
+
+void Component::_LoadDebugMesh() {
+    _DestroyDebugMesh();
+
+    if(mMeshHandle.isEmpty()) {
+        Logger::Get().Error("Component [" + mName + "]: Invalid handle for debug mesh.");
+    }
+
+    Ogre::SceneManager* scene_mgr = mNode->GetScene()->GetSceneManager();
+    std::string node_name = Utils::ToStdString(GetNode()->GetName());
```
Comment (svenstaro) [inline engine/src/Scene/Component.cpp:131]: Please don't use std::string as we use QString everywhere.

Comment (EternalWind) [inline engine/src/Scene/Component.cpp:131]: But the MeshComponent's _LoadMesh() method uses std::string as well.

File: engine/src/Scene/Component.cpp [hunk 6]
```diff
@@ -61,6 +68,8 @@ QScriptValue Component::GetScriptNode() {
 void Component::Create() {
     if(!mIsCreated) {
         mIsCreated = true;
+		_LoadDebugMesh();
```
Comment (svenstaro) [inline engine/src/Scene/Component.cpp]: Indentation

File: engine/src/Scene/Component.cpp (hunk 1 above)
Comment (svenstaro) [inline engine/src/Scene/Component.cpp]: Still indentation problem.

File: engine/src/Scene/Component.hpp [hunk 7]
```diff
@@ -148,6 +179,11 @@ protected:
     QString mName;  //!< The Component name.
     Node* mNode;        //!< The parent Node.
 
+    QString mMeshHandle;            //!< The handle of the mesh.
+    QString mMeshComponentName;     //!< The name of the mesh component.
```
Comment (svenstaro) [inline engine/src/Scene/Component.hpp]: Why do you need this? You never use it.

File: engine/src/Scene/Component.cpp (hunk 5 above)
Comment (svenstaro) [inline engine/src/Scene/Component.cpp:131]: Oh you are right. I just looked into that and it's actually the string that gets passed to Ogre and it is implicitly converted to Ogre's string. It's ok then.

**Comment by svenstaro** [comment]
Code looks good now. Are you going to add the actual meshes? If you need help with Blender, give me a shout on IRC.

**Comment by EternalWind** [comment]
Currently I'm thinking about making some changes on hiding/showing the debug shape. I got 2 ideas. One is moving the hide/show method to the node and make it can hide/show all the debug shape attached to the node. The other is adding a global switch that can control all the debug shape's visibility (no matter which node it is attached to). Both of the above will replace the current hide/show method because I think the current one is too inconvenient. Your thoughts? In addition, can you give me some ideas about some components' debug model? Like the ScriptComponent, the  FollowPathComponent and the PhysicsBodyComponent.

File: engine/src/Graphics/CameraComponent.hpp [hunk 8]
```diff
@@ -33,7 +33,7 @@ public:
       * @name The name of the component.
       * @see Component
       */
-    CameraComponent(const QString& name = "");
+    CameraComponent(const QString& name = "", const QString& mesh_handle_d = "camera.mesh");
```
Comment (svenstaro) [inline engine/src/Graphics/CameraComponent.hpp:37]: You didn't document mesh_handle_d here and it many other places.

File: engine/src/Graphics/LightComponent.hpp [hunk 9]
```diff
@@ -32,9 +32,10 @@ public:
     /**
       * Advanced constructor.
       * @param name The name for this component.
+      * @param mesh_handle The handle of the LightComponent's debug shape. Default is torch.mesh.
```
Comment (svenstaro) [inline engine/src/Graphics/LightComponent.hpp]: You say torch.mesh but you called it light.mesh?

**Comment by svenstaro** [comment]
Closing due to inactivity.

**Pull request closed without merging**